### PR TITLE
Add ListModel and ListStore bindings

### PR DESCRIPTION
**Pull request by sdroege**
This would later allow us to add the model support of `gtk::ListBox`

### CODE REVIEW

**Comment by sdroege** [comment]
CC @EPashkin 

Currently this fails to compile because of https://github.com/gtk-rs/sys/blob/b76cceca2bc81c822f1bb7320783824bdd0f79fe/gio-sys/src/lib.rs#L8240-L8241 . The reason for that is this in the GIR file

```xml
          <parameter name="additions" transfer-ownership="none">
            <doc xml:space="preserve">the items to add</doc>
            <array length="3" zero-terminated="0" c:type="gpointer*">
              <type name="GObject.Object"/>
            </array>
          </parameter>
```

Currently gir is checking for subclassing relationship between the types given there, however it does not consider raw pointers. This is in `codegen/sys/ffi_type.rs:ffi_inner()`. For any pointer-typed parameter, we should consider `gpointer` valid here.

I'm not sure how to teach gir that though, I only got it to produce `*mut gobject_ffi::GObject` (one `*mut` missing) as the information that this should be a pointer to a pointer seems to be missing at that place.

File: src/auto/list_store.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+pub trait ListStoreExt {
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn append<P: IsA<glib::Object>>(&self, item: &P);
+
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn insert<P: IsA<glib::Object>>(&self, position: u32, item: &P);
+
+    //#[cfg(any(feature = "v2_44", feature = "dox"))]
+    //fn insert_sorted<P: IsA<glib::Object>, Q: Into<Option</*Unimplemented*/Fundamental: Pointer>>>(&self, item: &P, compare_func: /*Unknown conversion*//*Unimplemented*/CompareDataFunc, user_data: Q) -> u32;
```
Comment (sdroege) [inline src/auto/list_store.rs]: For these scoped function pointers we should probably add some support to gir also, it's well-defined how they work

File: src/list_store.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn insert_sorted<P: IsA<glib::Object>, F: FnMut(&Object, &Object) -> Ordering>(&self, item: &P, compare_func: F) -> u32;
+
+    #[cfg(any(feature = "v2_46", feature = "dox"))]
+    fn sort<F: FnMut(&Object, &Object) -> Ordering>(&self, compare_func: F);
+}
+
+impl<O: IsA<ListStore> + IsA<glib::object::Object>> ListStoreExtManual for O {
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn insert_sorted<P: IsA<glib::Object>, F: FnMut(&Object, &Object) -> Ordering>(&self, item: &P, compare_func: F) -> u32 {
```
[the remark this answers is not in the document]
Comment (EPashkin) [reply src/list_store.rs]: You right

**Comment by EPashkin** [comment]
@sdroege You right about `g_list_store_splice`.
You add PR to gir?

**Comment by sdroege** [comment]
> @sdroege You right about g_list_store_splice.
> You add PR to gir?

I tried but I couldn't figure out how to fix this in gir, see https://github.com/gtk-rs/gio/pull/131#issuecomment-401591038

**Comment by EPashkin** [comment]
Ok, I try look at this.
Function need be
```
pub fn g_list_store_splice(store: *mut GListStore, position: c_uint, n_removals: c_uint,
additions: *mut [*mut GObject], n_additions: c_uint);
```
or
```
pub fn g_list_store_splice(store: *mut GListStore, position: c_uint, n_removals: c_uint,
additions: *mut [*mut gpointer], n_additions: c_uint);
```
?

**Comment by sdroege** [comment]
It needs to be `pub fn g_list_store_splice(store: *mut GListStore, position: c_uint, n_removals: c_uint,
additions: *mut *mut gobject_ffi::GObject, n_additions: c_uint);`

**Comment by EPashkin** [comment]
It really difficult problem, I currently no idea how to fix gir.
Maybe just add this function manual way?

**Comment by sdroege** [comment]
> It really difficult problem, I currently no idea how to fix gir.
> Maybe just add this function manual way?

Why are we doing these checks for matching types? If there is a type annotation like here, we should be able to always trust it.

And instead of manually implementing, I could also update `gir-files` to use the correct type there (`GObject**`) instead.

**Comment by EPashkin** [comment]
Don't know why we check it.
Agree that changing Gio-2.0.gir is better option

**Comment by sdroege** [comment]
> Agree that changing Gio-2.0.gir is better option

Will do that then tonight, thanks :)

**Comment by EPashkin** [comment]
Hm, seems solution found in gir, I missed that gpointer add "*mut" to base type itself
```
            if inner != glib_name {
+                if inner == "gpointer" {
+                    fix_name(env, tid, &glib_name).map_any(|s| format!("*mut {}", s))
                } else if implements_c_type(env, tid, &inner) {

```

**Comment by EPashkin** [comment]
@sdroege I restarted appveyor, it now fail on examples

**Comment by sdroege** [comment]
It duesnt seem to use the new version of sys

**Comment by EPashkin** [comment]
Its new error https://ci.appveyor.com/project/GuillaumeGomez/gio/build/1.0.437/job/prrgkvcjw07dxph5#L289
different from https://ci.appveyor.com/project/GuillaumeGomez/gio/build/1.0.436/job/fvtfjtjx9flbydhb#L276
so new sys used

**Comment by sdroege** [comment]
Ah i looked at Travis. Interesting!

**Comment by sdroege** [comment]
@EPashkin @GuillaumeGomez can you restart travis too? This looks weird

**Comment by GuillaumeGomez** [comment]
I restarted travis.

**Comment by EPashkin** [comment]
Restarting won't help, error repeats and it appears after this PR.

**Comment by sdroege** [comment]
> Restarting won't help, error repeats and it appears after this PR.

Travis didn't have this error though. It also makes no sense to me :) To you? Maybe `GtkListStore` vs `GListStore` confusion?

**Comment by EPashkin** [comment]
Travis have same error: https://travis-ci.org/gtk-rs/gio/jobs/398752719#L901
Yes, it ListStoreExtManual traits confusion, we need rename gtk's variants (Ext too) as we do with applicationExt
This PR IMHO ready for merge, as error solved by `use gtk::ListStoreExtManual;` or fixing gtk

**Comment by EPashkin** [comment]
Or we can rename gtk first then restart travis here to be sure

**Comment by sdroege** [comment]
I'll send gtk rename PR later tonight if nobody is faster

**Comment by GuillaumeGomez** [comment]
Done in https://github.com/gtk-rs/gtk/pull/670.

**Comment by EPashkin** [comment]
@GuillaumeGomez Please, restart CI here

**Comment by GuillaumeGomez** [comment]
Yes sir!

**Comment by sdroege** [comment]
All green! Thanks for your help :) Tomorrow I'll add the `GtkListBox` changes for this, and then an example

**Comment by sdroege** [comment]
@GuillaumeGomez good to go now :)

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**